### PR TITLE
ENH: DOC: improving error messages for casting errors in ufuncs #14828

### LINKED ISSUE #14828: Incorrect error message when adding i8 array to u8 array
When attempting to add an `ndarray` of type `int64` to an `ndarray` of type `uint64`, `numpy` throws a `TypeError`:

    TypeError: Cannot cast ufunc add output from dtype('float64') to dtype('uint64') with casting rule 'same_kind'

The error message incorrectly identifies the type of the first array as `float64`, when it is actually `int64`.

### Reproducing code example:

```python
import numpy as np
x = np.arange(5, dtype='u8')
y = np.arange(5, dtype='i8')
x += y
```

### Error message:

    TypeError: Cannot cast ufunc add output from dtype('float64') to dtype('uint64') with casting rule 'same_kind'

### Numpy/Python version information:

Tested with Python 3:

    1.16.4 3.6.8 (default, Oct  7 2019, 12:59:55) 
    [GCC 8.3.0]

Also with Python 2:

    ('1.15.1', '2.7.15+ (default, Oct  7 2019, 17:39:04) \n[GCC 7.4.0]')

### PR DESCRIPTION
**Pull request by micha2718l**
ENH: DOC: improving error messages for casting errors in ufuncs

This PR adds functionallity to create better error messages in the case that
casting can not be completed in application of ufuncs.
There are probably still some bugs or edge cases, as well as formatting issues;
but the functions seem to work to give messages such as:
`UFuncTypeError: Output of ufunc add(uint64, int64, out=uint64) resolved to the add(float64, float64, out=float64) loop, but np.can_cast(np.float64, np.uint64, casting='same_kind') is False`

Fixes #14828

### CODE REVIEW

**Comment by mattip** [comment]
In principle this looks like an improvement. Could you show a direct comparison of old-to-new? Also, is there a short-hand notation we could use to keep the error message length from getting out of hand, perhaps `dtype.str` would be more compact

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 1]
```diff
@@ -158,12 +160,28 @@ raise_casting_error(
...
     }
 
+    int j, nin = ufunc->nin, nop = nin + ufunc->nout;
+    PyObject *froms = PyTuple_New(nop);
+    PyObject *tos = PyTuple_New(nop);
+
+    if (input) {
+        for(j=0;j<nop;j++){
+            PyTuple_SetItem(froms, j, (PyObject *)PyArray_DESCR(operands[j]));
+            PyTuple_SetItem(tos, j, (PyObject *)dtypes[j]);
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c]: Missing `Py_INCREF(value)` here - `PyTuple_SetItem` steals references.

Also, you might as well use `PyTuple_SETITEM`, it's faster and no more dangerous

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 2]
```diff
@@ -158,12 +160,28 @@ raise_casting_error(
         return -1;
     }
 
+    int j, nin = ufunc->nin, nop = nin + ufunc->nout;
+    PyObject *froms = PyTuple_New(nop);
+    PyObject *tos = PyTuple_New(nop);
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c]: Need to handle the case when these return NULL. Simply giving up and returning the error would be fine.

**Comment by eric-wieser** [comment]
These lines will likely need to change too:
```
https://github.com/numpy/numpy/blob/dcc1fc2f055253703d3883ad186c26bcf6b9b276/numpy/core/_methods.py#L83-L96
```

**Comment by eric-wieser** [comment]
I think it might be worth renaming the python class members to be something like:
```
@_display_as_base
class _UFuncCastingError(UFuncTypeError):
    def __init__(self, ufunc, casting, call_dtypes, loop_dtypes):
        super().__init__(ufunc)
        self.casting = casting
        self.call_dtypes = call_dtypes
        self.loop_dtypes = loop_dtypes

@_display_as_base
class _UFuncInputCastingError(_UFuncCastingError):
    # other members as before
    @property
    def from_(self):
        return self.call_dtypes[self.in_i]
    @property
    def to(self):
        return self.loop_dtypes[self.in_i]

@_display_as_base
class _UFuncOutputCastingError(_UFuncCastingError):
    # other members as before
    @property
    def from_(self):
        return self.loop_dtypes[self.ufunc.nin + self.out_i]
    @property
    def to(self):
        return self.call_dtypes[self.ufunc.nin + self.out_i]
```

File: numpy/core/src/umath/ufunc_type_resolution.c (hunk 2 above)
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_type_resolution.c]: That is:
```C
PyObject *froms = PyTuple_New(nop);
if (froms == NULL) { 
    return NULL;
}
```
etc

File: numpy/core/src/umath/ufunc_type_resolution.c (hunk 1 above)
Comment (micha2718l) [reply numpy/core/src/umath/ufunc_type_resolution.c]: Thanks for all your helpful comments so far!
I included the `Py_INCREF` on the `dtypes[j]` object and it removed a segfault in an edge case I found. I don't think there needs to be one with regards to the `PyArray_DESCR(operands[j])` object because `PyArray_DESRC` returns a borrowed reference, correct me if i misunderstand something there.

Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_type_resolution.c]: Both are borrowed references, and this is exactly why you do need to incref them an extra time - when you call a function like `SETITEM` that steals a reference, you must make sure the reference is yours to give (ie not borrowed).

**Review by micha2718l — COMMENTED** [review]
@mattip 
> In principle this looks like an improvement. Could you show a direct comparison of old-to-new? Also, is there a short-hand notation we could use to keep the error message length from getting out of hand, perhaps `dtype.str` would be more compact

For the original (mentioned in issue #14843) example:
```
import numpy as np
x = np.arange(5, dtype='u8')
y = np.arange(5, dtype='i8')
x += y
```
The result for the current codebase is:
`TypeError: Cannot cast ufunc add output from dtype('float64') to dtype('uint64') with casting rule 'same_kind'`

The result from this change is:
`UFuncTypeError: Output of ufunc add(uint64, int64, out=uint64) resolved to the add(float64, float64, out=float64) loop, but np.can_cast(np.float64, np.uint64, casting='same_kind') is False`

I agree that that a shorter message is better, but I opted for the np.* form because it allows the code from the message to be directly copy and pasted into a REPL to play with it; if shorter takes precedence I can take the suggestion and make it less verbose.

**Comment by adeak** [comment]
As only a regular user of numpy the mixed (with respect to the presence of `np`) error message seems weird to me. The first two function calls and their arguments are without `np`, and I think it would look both shorter and better to omit them altogether:
>  UFuncTypeError: Output of ufunc add(uint64, int64, out=uint64) resolved to the add(float64, float64, out=float64) loop, but can_cast(float64, uint64, casting='same_kind') is False

It's true that you can't directly throw the last call into a REPL if you only have `numpy` imported `as np`, but you can alternatively import those three names instead. Or trust the interpreter when it says the result is `False` :)

**Comment by micha2718l** [comment]
> As only a regular user of numpy the mixed (with respect to the presence of `np`) error message seems a bit weird to me. The first two function calls and their arguments are without `np`, and I think it would look both shorter and better to omit them altogether:
> 
> > UFuncTypeError: Output of ufunc add(uint64, int64, out=uint64) resolved to the add(float64, float64, out=float64) loop, but can_cast(float64, uint64, casting='same_kind') is False
> 
> It's true that you can't directly throw the last call into a REPL if you only have `numpy` imported `as np`, but you can alternatively import those three names instead. Or trust the interpreter when it says the result is `False` :)

I totally see your point with the consistency. Any other opinions out there? If it were my choice I would want to have `np` on all parts, though it does create longer messages. I think there are valid points on both sides.

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 3]
```diff
@@ -158,12 +160,46 @@ raise_casting_error(
...
+            PyTuple_SET_ITEM(tos, j, (PyObject *)dtypes[j]);
+        }
+    }else{
+        for(j=0;j<nop;j++){
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(froms, j, (PyObject *)dtypes[j]);
+            if (operands[j] == NULL){
+                PyTuple_SET_ITEM(tos, j, Py_None);
+            } else {
+                Py_INCREF(operands[j]);
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:191]: This should incref the dtype, not the operand

File: numpy/core/_exceptions.py [hunk 4]
```diff
@@ -106,13 +116,23 @@ def __init__(self, ufunc, casting, from_, to, i):
         self.out_i = i
 
     def __str__(self):
-        # only show the number if more than one output exists
-        i_str = "{} ".format(self.out_i) if self.ufunc.nout != 1 else ""
+        if self.ufunc.nout > 1:
+            from_outs = ", ".join(["out{}={}".format(i+1, f.name) for i, f in enumerate(self.from_) if i>=self.ufunc.nin])
+            to_outs = ", ".join(["out{}={}".format(i+1, t.name) for i, t in enumerate(self.to) if i>=self.ufunc.nin])
```
Comment (eric-wieser) [inline numpy/core/_exceptions.py:121]: This doesn't produce a legal call, it should be:
```suggestion
            to_outs = "out=({})".format(", ".join(t.name for t in self.to[self.ufunc.nin:]))
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 5]
```diff
@@ -158,12 +160,46 @@ raise_casting_error(
...
+            if (operands[j] == NULL){
+                PyTuple_SET_ITEM(froms, j, Py_None);
+            } else {
+                Py_INCREF(operands[j]);
+                PyTuple_SET_ITEM(froms, j, (PyObject *)PyArray_DESCR(operands[j]));
+            }
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(tos, j, (PyObject *)dtypes[j]);
+        }
+    }else{
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:184]: Style:
```suggestion
    }
    else {
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 6]
```diff
@@ -158,12 +160,46 @@ raise_casting_error(
...
+    if (froms == NULL) { 
+        return -1;
+    }
+    PyObject *tos = PyTuple_New(nop);
+    if (tos == NULL) { 
+        return -1;
+    }
+
+    if (input) {
+        for(j=0;j<nop;j++){
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:174]:
```suggestion
        for(int j=0; j < nop; j++){
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 7]
```diff
@@ -158,12 +160,46 @@ raise_casting_error(
...
+            if (operands[j] == NULL){
+                PyTuple_SET_ITEM(froms, j, Py_None);
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:176]: Need to incref None here

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 8]
```diff
@@ -158,12 +160,46 @@ raise_casting_error(
...
+            }
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(tos, j, (PyObject *)dtypes[j]);
+        }
+    }else{
+        for(j=0;j<nop;j++){
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(froms, j, (PyObject *)dtypes[j]);
+            if (operands[j] == NULL){
+                PyTuple_SET_ITEM(tos, j, Py_None);
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:189]: Needs to incref None

**Comment by eric-wieser** [comment]
> Any other opinions out there? If it were my choice I would want to have `np` on all parts

Numpy is not the only module capable of defining ufuncs, so prepending `np.` before `add` would most likely be incorrect for errors from, say, scipy ufuncs

**Comment by eric-wieser** [comment]
@micha2718l, do you think you'll have time to come back to this?

File: numpy/core/_exceptions.py [hunk 9]
```diff
@@ -88,13 +88,23 @@ def __init__(self, ufunc, casting, from_, to, i):
...
+            "Input of ufunc {}({}, {}) resolved to the {}({}, {}) loop," 
+            " but np.can_cast(np.{}, np.{}, casting='{}') is False"
+        ).format(self.ufunc.__name__,
+                 ", ".join([f.name for i, f in enumerate(self.to) if i<self.ufunc.nin]),
+                 to_outs,
+                 self.ufunc.__name__,
+                 ", ".join([f.name for i, f in enumerate(self.from_) if i<self.ufunc.nin]),
+                 from_outs,
+                 self.from_[self.in_i].name, self.to[self.in_i].name, self.casting
         )
```
Comment (eric-wieser) [inline numpy/core/_exceptions.py:108]:
```suggestion
        ).format(
            self.ufunc.__name__,
            ", ".join([f.name for i, f in enumerate(self.to) if i<self.ufunc.nin]),
            to_outs,
            self.ufunc.__name__,
            ", ".join([f.name for i, f in enumerate(self.from_) if i<self.ufunc.nin]),
            from_outs,
            self.from_[self.in_i].name, self.to[self.in_i].name, self.casting
        )
```

File: numpy/core/_exceptions.py [hunk 10]
```diff
@@ -106,13 +116,23 @@ def __init__(self, ufunc, casting, from_, to, i):
...
+            "Output of ufunc {}({}, {}) resolved to the {}({}, {}) loop," 
+            " but np.can_cast(np.{}, np.{}, casting='{}') is False"
+        ).format(self.ufunc.__name__,
+                 ", ".join([f.name for i, f in enumerate(self.to) if i<self.ufunc.nin]),
+                 to_outs,
+                 self.ufunc.__name__,
+                 ", ".join([f.name for i, f in enumerate(self.from_) if i<self.ufunc.nin]),
+                 from_outs,
+                 self.from_[self.out_i].name, self.to[self.out_i].name, self.casting
         )
```
Comment (eric-wieser) [inline numpy/core/_exceptions.py:136]:
```suggestion
        ).format(
            self.ufunc.__name__,
            ", ".join([f.name for i, f in enumerate(self.to) if i<self.ufunc.nin]),
            to_outs,
            self.ufunc.__name__,
            ", ".join([f.name for i, f in enumerate(self.from_) if i<self.ufunc.nin]),
            from_outs,
            self.from_[self.out_i].name, self.to[self.out_i].name, self.casting
        )
```

File: numpy/core/src/umath/ufunc_type_resolution.c (hunk 7 above)
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:176]:
```suggestion
            if (operands[j] == NULL) {
                Py_INCREF(Py_None);
                PyTuple_SET_ITEM(froms, j, Py_None);
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 11]
```diff
@@ -158,12 +160,46 @@ raise_casting_error(
...
+    PyObject *tos = PyTuple_New(nop);
+    if (tos == NULL) { 
+        return -1;
+    }
+
+    if (input) {
+        for(j=0;j<nop;j++){
+            if (operands[j] == NULL){
+                PyTuple_SET_ITEM(froms, j, Py_None);
+            } else {
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:177]:
```suggestion
            }
            else {
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 12]
```diff
@@ -158,12 +160,46 @@ raise_casting_error(
...
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(tos, j, (PyObject *)dtypes[j]);
+        }
+    }else{
+        for(j=0;j<nop;j++){
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(froms, j, (PyObject *)dtypes[j]);
+            if (operands[j] == NULL){
+                PyTuple_SET_ITEM(tos, j, Py_None);
+            } else {
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:190]:
```suggestion
            }
            else {
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 13]
```diff
@@ -158,12 +160,46 @@ raise_casting_error(
...
+                PyTuple_SET_ITEM(froms, j, (PyObject *)PyArray_DESCR(operands[j]));
+            }
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(tos, j, (PyObject *)dtypes[j]);
+        }
+    }else{
+        for(j=0;j<nop;j++){
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(froms, j, (PyObject *)dtypes[j]);
+            if (operands[j] == NULL){
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:188]:
```suggestion
            if (operands[j] == NULL) {
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 14]
```diff
@@ -158,12 +160,46 @@ raise_casting_error(
...
+        return -1;
+    }
+    PyObject *tos = PyTuple_New(nop);
+    if (tos == NULL) { 
+        return -1;
+    }
+
+    if (input) {
+        for(j=0;j<nop;j++){
+            if (operands[j] == NULL){
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:175]:
```suggestion
            if (operands[j] == NULL) {
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 15]
```diff
@@ -158,12 +160,49 @@ raise_casting_error(
...
+            else {
+                Py_INCREF(operands[j]);
+                PyTuple_SET_ITEM(froms, j, (PyObject *)PyArray_DESCR(operands[j]));
+            }
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(tos, j, (PyObject *)dtypes[j]);
+        }
+    }
+    else {
+        for(j=0;j<nop;j++){
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:187]:
```suggestion
        for(int j=0; j<nop; j++) {
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 16]
```diff
@@ -146,10 +146,12 @@ raise_casting_error(
...
         NPY_CASTING casting,
-        PyArray_Descr *from,
-        PyArray_Descr *to,
-        npy_intp i)
+        PyArrayObject **operands,
+        PyArray_Descr **dtypes,
+        npy_intp i,
+        bool input)
 {
+
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:154]:
```suggestion
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 17]
```diff
@@ -158,12 +160,49 @@ raise_casting_error(
         return -1;
     }
 
+    int j, nin = ufunc->nin, nop = nin + ufunc->nout;
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:163]:
```suggestion
    int nin = ufunc->nin;
    int nop = nin + ufunc->nout;
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 18]
```diff
@@ -158,12 +160,49 @@ raise_casting_error(
...
+    if (froms == NULL) { 
+        return -1;
+    }
+    PyObject *tos = PyTuple_New(nop);
+    if (tos == NULL) { 
+        return -1;
+    }
+
+    if (input) {
+        for(int j=0; j < nop; j++) {
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:174]:
```suggestion
        for(int j = 0; j < nop; j++) {
```

File: numpy/core/src/umath/ufunc_type_resolution.c (hunk 8 above)
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_type_resolution.c:191]:
```suggestion
                Py_INCREF(Py_None);
                PyTuple_SET_ITEM(tos, j, Py_None);
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 19]
```diff
@@ -158,12 +160,49 @@ raise_casting_error(
...
+    else {
+        for(j=0;j<nop;j++){
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(froms, j, (PyObject *)dtypes[j]);
+            if (operands[j] == NULL) {
+                PyTuple_SET_ITEM(tos, j, Py_None);
+            }
+            else {
+                Py_INCREF(operands[j]);
+                PyTuple_SET_ITEM(tos, j, (PyObject *)PyArray_DESCR(operands[j]));
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:195]:
```suggestion
                PyObject *dtype = (PyObject *)PyArray_DESCR(operands[j]);
                Py_INCREF(dtype);
                PyTuple_SET_ITEM(tos, j, dtype);
```

File: numpy/core/_exceptions.py [hunk 20]
```diff
@@ -88,13 +88,24 @@ def __init__(self, ufunc, casting, from_, to, i):
...
+            to_outs = ", ".join(["out{}={}".format(i+1, t.name) for i, t in enumerate(self.to) if i>=self.ufunc.nin])
+        else:
+            from_outs = "out={}".format(self.from_[-1])
+            to_outs = "out={}".format(self.to[-1])
+            
         return (
-            "Cannot cast ufunc {!r} input {}from {!r} to {!r} with casting "
-            "rule {!r}"
+            "Input of ufunc {}({}, {}) resolved to the {}({}, {}) loop," 
+            " but np.can_cast(np.{}, np.{}, casting='{}') is False"
```
Comment (eric-wieser) [inline numpy/core/_exceptions.py:100]:
```suggestion
            "Input of ufunc {}({}, {}) resolved to the {}({}, {}) loop, " 
            "but np.can_cast(np.{}, np.{}, casting='{}') is False"
```

File: numpy/core/_exceptions.py [hunk 21]
```diff
@@ -106,13 +117,24 @@ def __init__(self, ufunc, casting, from_, to, i):
...
+            to_outs = ", ".join(["out{}={}".format(i+1, t.name) for i, t in enumerate(self.to) if i>=self.ufunc.nin])
+        else:
+            from_outs = "out={}".format(self.from_[-1])
+            to_outs = "out={}".format(self.to[-1])
+            
         return (
-            "Cannot cast ufunc {!r} output {}from {!r} to {!r} with casting "
-            "rule {!r}"
+            "Output of ufunc {}({}, {}) resolved to the {}({}, {}) loop," 
+            " but np.can_cast(np.{}, np.{}, casting='{}') is False"
```
Comment (eric-wieser) [inline numpy/core/_exceptions.py:129]:
```suggestion
            "Output of ufunc {}({}, {}) resolved to the {}({}, {}) loop, " 
            "but np.can_cast(np.{}, np.{}, casting='{}') is False"
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 22]
```diff
@@ -158,12 +159,50 @@ raise_casting_error(
...
+            else {
+                Py_INCREF(operands[j]);
+                PyTuple_SET_ITEM(froms, j, (PyObject *)PyArray_DESCR(operands[j]));
+            }
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(tos, j, (PyObject *)dtypes[j]);
+        }
+    }
+    else {
+        for(int j = 0; j < nop; j++) {
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:187]:
```suggestion
        for (int j = 0; j < nop; j++) {
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 23]
```diff
@@ -158,12 +159,50 @@ raise_casting_error(
...
+    if (froms == NULL) { 
+        return -1;
+    }
+    PyObject *tos = PyTuple_New(nop);
+    if (tos == NULL) { 
+        return -1;
+    }
+
+    if (input) {
+        for(int j = 0; j < nop; j++) {
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:174]:
```suggestion
        for (int j = 0; j < nop; j++) {
```

File: numpy/core/_exceptions.py (hunk 4 above)
Comment (eric-wieser) [reply numpy/core/_exceptions.py:122]: (same for the other three lines)

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 24]
```diff
@@ -158,12 +159,50 @@ raise_casting_error(
...
+            PyTuple_SET_ITEM(froms, j, (PyObject *)dtypes[j]);
+            if (operands[j] == NULL) {
+                PyTuple_SET_ITEM(tos, j, Py_None);
+            }
+            else {
+                Py_INCREF(operands[j]);
+                PyTuple_SET_ITEM(tos, j, (PyObject *)PyArray_DESCR(operands[j]));
+            }
+        }
+    }
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:198]: Let's remove some duplication here:
```suggestion
    PyObject *op_dtype_tup = PyTuple_New(nop);
    if (froms == NULL) { 
        return -1;
    }
    PyObject *dtype_tup = PyTuple_New(nop);
    if (tos == NULL) { 
        return -1;
    }

    for (int j = 0; j < nop; j++) {
        if (operands[j] == NULL) {
            PyTuple_SET_ITEM(op_dtype_tup, j, Py_None);
        }
        else {
            Py_INCREF(operands[j]);
            PyTuple_SET_ITEM(op_dtype_tup, j, (PyObject *)PyArray_DESCR(operands[j]));
        }
        Py_INCREF(dtypes[j]);
        PyTuple_SET_ITEM(dtype_tup, j, (PyObject *)dtypes[j]);
    }
    
    PyObject *froms = input ? op_dtype_tup : dtype_tup;
    PyObject *tos = input ? dtype_tup :  op_dtype_tup;
```

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 25]
```diff
@@ -172,15 +172,23 @@ raise_casting_error(
 
     if (input) {
         for(j=0;j<nop;j++){
-            PyTuple_SET_ITEM(froms, j, (PyObject *)PyArray_DESCR(operands[j]));
+            if (operands[j] == NULL){
+                PyTuple_SET_ITEM(froms, j, Py_None);
+            } else {
+                PyTuple_SET_ITEM(froms, j, (PyObject *)PyArray_DESCR(operands[j]));
+            }
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_type_resolution.c:181]: Good question, I don't know the answer

File: numpy/core/src/umath/ufunc_type_resolution.c (hunk 19 above)
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_type_resolution.c:195]: Moved to https://github.com/numpy/numpy/pull/14843/files#r412110682

File: numpy/core/src/umath/ufunc_type_resolution.c (hunk 8 above)
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_type_resolution.c:191]: https://github.com/numpy/numpy/pull/14843/files#r412110682

File: numpy/core/src/umath/ufunc_type_resolution.c (hunk 7 above)
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_type_resolution.c:176]: Moved to https://github.com/numpy/numpy/pull/14843/files#r412110682

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 26]
```diff
@@ -158,12 +159,50 @@ raise_casting_error(
...
+            else {
+                Py_INCREF(operands[j]);
+                PyTuple_SET_ITEM(tos, j, (PyObject *)PyArray_DESCR(operands[j]));
+            }
+        }
+    }
+
     exc_value = Py_BuildValue(
         "ONOOi",
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c:201]:
```suggestion
        "ONNNi",
```
These are new references, so you either need to decref them after the call, or pass the ownership to the tuple.

**Comment by micha2718l** [comment]
> @micha2718l, do you think you'll have time to come back to this?
@eric-wieser 
Yes, especially now with all the free time from quarantine. I'll swing back to this in the next couple days for sure.

**Comment by mattip** [comment]
@micha2718l ping

**Comment by mattip** [comment]
@micha2718l it would be nice to finish this up. We are nearing the cutoff for the next release.

File: numpy/core/_exceptions.py (hunk 4 above)
Comment (eric-wieser) [reply numpy/core/_exceptions.py:122]: This comment still applies

**Comment by micha2718l** [comment]
@mattip @eric-wieser 
I *think* I took care of the outstanding issues with this PR. Let me know if there is any other details to think about.

**Comment by adeak** [comment]
I only have the same trivial remark as earlier: the mixture of `np.can_cast` and `np.`/non-`np.` types in that error message look a bit weird to me :)

But this is might be bikeshedding territory, and I don't have the perspective to tell whether my preference is actuallybetter (but unless I missed it I don't think we got a clear core dev statement on the matter, beyond that `np.add` would definitely be off). I'm sorry if I just missed it.

**Comment by micha2718l** [comment]
> I only have the same trivial remark as earlier: the mixture of `np.can_cast` and `np.`/non-`np.` types in that error message look a bit weird to me :)
> 
> But this is might be bikeshedding territory, and I don't have the perspective to tell whether my preference is actually better (but unless I missed it I don't think we got a clear core dev statement on the matter, beyond that `np.add` would definitely be off). I'm sorry if I just missed it.

I missed that comment when I returned to this (and it had been a while so I had forgot about your comment). I still think there are valid points on either side, if any core have an opinion it is welcome. I'm open to any of the three potential options here, all np, no np, or the hybrid as is; which balances brevity and usefulness IMHO

**Comment by eric-wieser** [comment]
Thanks for coming back to this! I think this comment still applies: https://github.com/numpy/numpy/pull/14843#discussion_r402719113

File: numpy/core/_exceptions.py [hunk 27]
```diff
@@ -88,13 +88,24 @@ def __init__(self, ufunc, casting, from_, to, i):
...
+            to_outs = ", ".join(["out{}={}".format(i+1, t.name) for i, t in enumerate(self.to) if i>=self.ufunc.nin])
+        else:
+            from_outs = "out={}".format(self.from_[-1])
+            to_outs = "out={}".format(self.to[-1])
+            
         return (
-            "Cannot cast ufunc {!r} input {}from {!r} to {!r} with casting "
-            "rule {!r}"
+            "Input of ufunc {}({}, {}) resolved to the {}({}, {}) loop, " 
+            "but np.can_cast(np.{}, np.{}, casting='{}') is False"
```
Comment (eric-wieser) [inline numpy/core/_exceptions.py:100]: I think I agree with @adeak, let's drop the `np.` prefixes

File: numpy/core/_exceptions.py [hunk 28]
```diff
@@ -88,13 +88,24 @@ def __init__(self, ufunc, casting, from_, to, i):
         self.in_i = i
 
     def __str__(self):
-        # only show the number if more than one input exists
-        i_str = "{} ".format(self.in_i) if self.ufunc.nin != 1 else ""
+        if self.ufunc.nout > 1:
+            from_outs = ", ".join(["out{}={}".format(i+1, f.name) for i, f in enumerate(self.from_) if i>=self.ufunc.nin])
+            to_outs = ", ".join(["out{}={}".format(i+1, t.name) for i, t in enumerate(self.to) if i>=self.ufunc.nin])
```
Comment (eric-wieser) [inline numpy/core/_exceptions.py:93]: To be clear - this `out{}=` needs changing to match the way you updated it for the other one.

Comment (micha2718l) [reply numpy/core/_exceptions.py:93]: Thanks, coming back into it I had forgotten there were two places where this applied.

**Comment by eric-wieser** [comment]
What are your thoughts on this comment: https://github.com/numpy/numpy/pull/14843#issuecomment-550566562?

File: numpy/core/_exceptions.py [hunk 29]
```diff
@@ -106,13 +117,24 @@ def __init__(self, ufunc, casting, from_, to, i):
...
-        # only show the number if more than one output exists
-        i_str = "{} ".format(self.out_i) if self.ufunc.nout != 1 else ""
+        if self.ufunc.nout > 1:
+            from_outs = "out=({})".format(", ".join(f.name for f in self.to[self.ufunc.nin:]))
+            to_outs = "out=({})".format(", ".join(t.name for t in self.to[self.ufunc.nin:]))
+        else:
+            from_outs = "out={}".format(self.from_[-1])
+            to_outs = "out={}".format(self.to[-1])
+
         return (
```
Comment (eric-wieser) [inline numpy/core/_exceptions.py:127]: Might be worth a helper function somewhere,
```python
def _ufunc_call_sig(ufunc, arg_dtypes):
    if ufunc.nout > 1:
        outs = "out=({})".format(", ".join(f.name for f in arg_dtypes[ufunc.nin:]))
    else:
        outs = "out={}".format(arg_dtypes[-1])
    return f"{}({}, {})".format(ufunc.__name__, ", ".join(arg_dtypes[:ufunc.nin]), outs)
```
which you can then call as `_ufunc_call_sig(self.ufunc, self.from)` and `_ufunc_call_sig(self.ufunc, self.to)` in the error messages

Comment (micha2718l) [reply numpy/core/_exceptions.py:127]: I do like this concept, I'm putting it together now to see how that works out. Helper function likely good, but either way something to have the meanings of from/to not switching around is a good thing.

Comment (eric-wieser) [reply numpy/core/_exceptions.py:127]: > I'm putting it together now to see how that works out.

Were you able to finish this up?

**Comment by mattip** [comment]
@micha2718l this has developed a conflict in the meantime. For other reviewers, the bottom line is that this example
```
import numpy as np
x = np.arange(5, dtype='u8')
y = np.arange(5, dtype='i8')
x += y
```

Currently displays
```
Cannot cast ufunc 'add' output from dtype('float64') to dtype('uint64') with casting rule 'same_kind'
```

and after this PR displays
```
Output of ufunc add(float64, float64, out=uint64) resolved to the add(float64, float64, out=float64) loop, but can_cast(float64, uint64, casting='same_kind') is False
```

What do other reviewers think, is this worth pushing forward? It seems a bit wordy to me.

**Comment by rossbar** [comment]
> What do other reviewers think, is this worth pushing forward? It seems a bit wordy to me.

My two cents: I agree with @mattip , I personally find the current wording more readable.

**Comment by mattip** [comment]
Maybe going back to the issue #14828 and reaching agreement around a desired error message should be a first step before the PR implementation.

**Comment by mattip** [comment]
Closing. Thanks for giving it a try, maybe once we have an idea what we want you can try again.
